### ISSUE TITLE
selectOne and selectExactlyOne: type incompatibility with pg's values interpolation

### ISSUE THREAD

**Issue by mrattner** [jawj/zapatos#84]
The generated SQL for `selectOne` and `selectExactlyOne` includes a `FETCH FIRST $n ROWS ONLY` clause, where $n is set to the value `1`. Example:

    const query = db.selectOne(
        'Evse',
        { uid: evseUid },
        { lateral: { connectors: db.select('Connector', { evse_id: db.parent('id') }) } },
    );

Compiled query text and values:

    {
        "text": "SELECT to_jsonb(\"Evse\".*) || jsonb_build_object($1::text, \"ljoin_0\".result) AS result FROM \"Evse\" LEFT JOIN LATERAL (SELECT coalesce(jsonb_agg(result), '[]') AS result FROM (SELECT to_jsonb(\"Connector\".*) AS result FROM \"Connector\" WHERE (\"evse_id\" = \"Evse\".\"id\")) AS \"sq_Connector\") AS \"ljoin_0\" ON true WHERE (\"uid\" = $2) FETCH FIRST $3 ROWS ONLY",
        "values": ["connectors","59b19e1e-46ac-4999-b111-6b42f0661dc1",1]
    }

### Steps to reproduce

Here's a [minimal database schema and example](https://gist.github.com/mrattner/1b8411aad236f71e1ef3e1761702d6c6) of executing the above query. `pg` version: 8.6.0, Postgres server version: 10.2.

### Result

    syntax error at or near "$3"
    length: 92,
    severity: 'ERROR',
    code: '42601',
    position: '337',
    file: 'scan.l',
    line: '1087',
    routine: 'scanner_yyerror'

This happens because when `pg` interpolates the values using its [`prepareValue` function](https://github.com/brianc/node-postgres/blob/6ced5243901a31a4030af13337ef2735c3eaaa9b/lib/utils.js), there is no special handling for values with the Javascript type `number`. Makes sense, since there isn't a clear mapping from a Javascript number to a particular Postgres number type, and if you care about the type, you can use `::` operator to cast it to the type you intend.

### Suggested fix
When generating the query text, add the type for parameters whose numeric types are known: e.g. change `FETCH FIRST $3 ROWS ONLY` to `FETCH FIRST ($3::smallint) ROWS ONLY` instead, which gives the expected result. File: [shortcuts.ts
](https://github.com/jawj/zapatos/blob/d88923219ee7012101c2e30ba303a8fb438b4cbc/src/db/shortcuts.ts)

### Workaround
Use the `select` or `sql` functions instead of selectOne/selectExactlyOne.

----

Perhaps this syntax error does not occur on more recent Postgres server versions? 10.2 is somewhat old, but I could not find out whether the syntax has changed in recent years such that `FETCH FIRST '1'` is now equivalent to `FETCH FIRST 1`. I'm curious whether that's the case, since I'm a bit of a postgres noob :blush: 

:+1: Thanks for making this great library available for public use!

**Comment by jawj** [comment]
Thanks for taking the time to post such a thorough issue.

I've not so far managed to reproduce this problem on the version of 10 I have installed, which is 10.16.

For example, this works:

```
db:query (-) SELECT to_jsonb("files".*) AS result FROM "files" FETCH FIRST $1 ROWS ONLY
db:query [ 1 ] +0ms
db:result (-, 0.5ms) {
db:result   id: 7,
db:result   path: '/otherimgs/a.jpg',
db:result   created_at: '1989-01-21T00:00:00+00:00',
db:result   updated_at: '2021-05-11T13:20:36.418+01:00'
db:result } +1ms
```

Before we go further, I note that in your minimal repro schema and example, the limit value placeholder is `$3` but only two values are supplied in the values array. Is this the values array Zapatos produces?

**Comment by mrattner** [comment]
No, I must have copy-pasted it wrong! Zapatos correctly produces 1 as the third value. I've updated the gist.
